contrib/memory/src/memory.nit

```diff
@@ -0,0 +1,994 @@
+					cpt += 1
+					setpath
+				else if cpt < level.length then
+					# The cursor is playing
+					var b = level[cpt]
+					b.ttl = 1.0
+					b.sound.play
+					cpt += 1
+					setpath
+				else if cpt == level.length then
+					# The cursor is out, run the shuffle
+					path = null
+					if shuffling == 1 then
+						shuffle(display)
+					else if shuffling > 1 then
+						hard_shuffle(display)
+					end
+					cpt += 1
+				else
+					# The shuffling is over, start playing
+					player = true
+					cpt = 0
+				end
+			end
+		else if ttl > 0.0 then
+			ttl -= 0.1
+			if ttl <= 0.0 then
+				ttl = 0.0
+				if cpt == level.length then
+					fly_out(display)
+					cpt += 1
+				else
+					if is_menu then
+						new_game
+					else if is_win then
+						current_level += 1
+						new_game
+					else
+						replay_game
+					end
+				end
+			end
+		end
+
+		# Display each button
+		for b in buttons do
+			b.blit_on(display)
+		end
+
+		# Display the cursor
+		if path != null then
+			drawing.cursor.scale = scale
+			display.blit(drawing.cursor, cursor.x, cursor.y)
+		end
+
+		if not is_menu then
+			blit_number(current_level, ui_scale, 10.0 * scale, 10.0 * scale)
+			reload.blit_on(display)
+		end
+	end
+
+	# Blit a number somewhere
+	fun blit_number(number: Int, scale: Float, x, y: Float)
+	do
+		for img in number_images.imgs do img.scale = scale
+		display.blit_number(number_images, number, x.to_i, y.to_i)
+	end
+
+	# Images with the numbers
+	private var number_images: NumberImages is noautoinit
+
+	# A player click on a button
+	fun action(b: Button)
+	do
+		if is_menu then
+			b.sound.play
+			mode = b.place
+			current_level = 0
+			ttl = 0.1
+			cpt = level.length
+			is_win = true
+			return
+		end
+		if cpt >= level.length then return
+		if b == level[cpt] then
+			b.sound.play
+			b.ttl = 1.0
+
+			cpt += 1
+			if cpt >= level.length then
+				is_win = true
+				print "Won!"
+				ttl = 2.0
+			end
+		else
+			error += 1
+			print "Err {error}"
+			b.error = true
+			b.ttl = 3.0
+			snd_penalty.play
+			if error > 2 then
+				is_win = false
+				print "Lose!"
+				for b2 in buttons do
+					b2.error = true
+					b2.ttl = 3.0
+				end
+				ttl = 3.0
+				cpt = level.length
+			end
+		end
+	end
+
+	redef fun input(ie)
+	do
+		# Quit?
+		if ie isa QuitEvent then
+			quit = true
+			return true
+		end
+
+		# On click (or tap)
+		if ie isa PointerEvent and ie.depressed then
+			if player then
+				for b in buttons do
+					if b.has(ie.x, ie.y) then
+						action(b)
+						return true
+					end
+				end
+			end
+
+			if not is_menu then
+				if reload.has(ie.x, ie.y) then
+					reload.sound.play
+					reload.ttl = 1.0
+					replay_game
+					return true
+				end
+			end
+		end
+
+		# Special commands
+		if ie isa KeyEvent and ie.is_down then
+			var c = ie.name
+
+			if c == "4" or c == "escape" then
+				if is_menu then
+					# quit = true # broken
+					new_menu
+				else
+					new_menu
+				end
+				return true
+			end
+			print c
```

Comment:
If this is a debug print you should add some context to the print for it to be helpful later.